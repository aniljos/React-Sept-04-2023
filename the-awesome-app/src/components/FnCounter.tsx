import {ChangeEvent, useState, useRef, useEffect} from 'react';

type FnCounterProps = {
    initValue?: number
}
function FnCounter( props: FnCounterProps){

    const [counter, setCounter] = useState(props.initValue ? props.initValue : 0);
    const inputRef = useRef<HTMLInputElement>(null);
    //let incrementCount = 0;
    const incrementCount = useRef(0);


    //signatue: useEffect(callback, [dependencies])
    // useEffect equivalent to componentDidMount => useEffect(callback, [])
    useEffect(() => {

        console.log("useEffect equivalent to componentDidMount...")

        // useEffect equivalent to componentWillUnMount 
                            //=> callback returned form the useEffect with no dependencies
        return () => {
            console.log("useEffect equivalent to componentWillUnMount...")
        }
    }, [])

    // useEffect equivalent to componentDidUpdate => useEffect(callback, [dependency])
    useEffect(()=> {
        console.log("useEffect equivalent to componentDidUpdate...", counter)
    }, [counter])

    function inc(){
        setCounter(counter + 1);
        console.log("Counter", counter);
        incrementCount.current++;
        console.log("incrementCount", incrementCount.current);

    }

    const decr = ()=> {
        setCounter(counter - 1);
    }

    function handleChange(evt: ChangeEvent<HTMLInputElement>){

        if(evt.target.value){
            setCounter(Number(evt.target.value))
        }
        else{
            setCounter(0)
        }
        
    }
    function handleUpdate(){
        //setCounter()
        console.log("inputRef: ", inputRef);
        console.log("inputRef: ", inputRef.current?.value);
        setCounter(Number(inputRef.current?.value))
    }

    return (
        <div>
            <h4>Fn Counter: {counter}</h4>
            <p>This is a functional component with hooks</p>
            <div>
                <button onClick={inc}>Inc</button> &nbsp;
                <button onClick={decr}>Decr</button>
            </div>
            <div>
                {/* Controlled Input */}
                Counter: <input type='number' 
                            placeholder='Counter' value={counter} onChange={handleChange}/>
            </div>
            <div>
                {/* Uncontrolled Input */}
                Update Counter: <input ref={inputRef} type='number' placeholder='Enter a value'/>&nbsp;
                <button onClick={handleUpdate}>Update</button>
            </div>
        </div>
    )
}

export default FnCounter