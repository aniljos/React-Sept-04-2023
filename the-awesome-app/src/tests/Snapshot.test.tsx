import { render } from "@testing-library/react";
import FnCounter from "../components/FnCounter";


//When the test is executed the first time it creates the shapshot
//To check change the init value to 9 or some other value, the test should fail
// To update the shapshots npm test -- --updateSnapshot
test('MyComponent snapshot', () => {
    const { asFragment } = render(<FnCounter initValue={8}  />);
    expect(asFragment()).toMatchSnapshot();
});