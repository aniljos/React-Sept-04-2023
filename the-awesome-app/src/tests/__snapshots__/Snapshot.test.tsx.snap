// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent snapshot 1`] = `
<DocumentFragment>
  <div>
    <h4>
      Fn Counter: 8
    </h4>
    <p>
      This is a functional component with hooks
    </p>
    <div>
      <button>
        Inc
      </button>
        
      <button>
        Decr
      </button>
    </div>
    <div>
      Counter: 
      <input
        placeholder="Counter"
        type="number"
        value="8"
      />
    </div>
    <div>
      Update Counter: 
      <input
        placeholder="Enter a value"
        type="number"
      />
       
      <button>
        Update
      </button>
    </div>
  </div>
</DocumentFragment>
`;
